lofire-repo/src/tree.rs

```diff
@@ -1 +1,555 @@
+            let data_chunk = ObjectContentV0::DataChunk(chunk.to_vec());
+            let content_ser = serde_bare::to_vec(&data_chunk).unwrap();
+            nodes.push(make_object(
+                content_ser.as_slice(),
+                &conv_key,
+                &vec![],
+                &ObjectDeps::ObjectIdList(vec![]),
+                expiry,
+            ));
+        }
+
+        // internal nodes
+        // arity: max number of ObjectRefs that fit inside an InternalNode Object within the object_size limit
+        let max_deps_size = 8 * OBJECT_ID_SIZE;
+        let arity: usize =
+            (object_size - EMPTY_OBJECT_SIZE - max_deps_size) / (OBJECT_ID_SIZE + OBJECT_KEY_SIZE);
+        let deps = make_deps(root_deps, object_size, repo_pubkey, repo_secret);
+        let mut parents = make_tree(nodes.as_slice(), &conv_key, &deps, expiry, arity);
+        nodes.append(&mut parents);
+
+        // root node
+        let (root_obj, root_key) = nodes.last().unwrap();
+        let root_id = Self::object_id(root_obj);
+
+        Tree {
+            root_id,
+            root_key: Some(root_key.clone()),
+            nodes: nodes.into_iter().map(|(obj, _key)| obj).collect(),
+        }
+    }
+
+    /// Load tree from store
+    ///
+    /// Returns Ok(Tree) or a Err(Vec<ObjectId>) of missing Object IDs
+    pub fn load(
+        store: &Store,
+        root_id: ObjectId,
+        root_key: Option<SymKey>,
+    ) -> Result<Tree, Vec<ObjectId>> {
+        fn load_tree(
+            nodes: &mut Vec<Object>,
+            missing: &mut Vec<ObjectId>,
+            store: &Store,
+            id: &ObjectId,
+        ) {
+            match store.get(id) {
+                Ok(obj) => {
+                    nodes.insert(0, obj.clone());
+                    match obj {
+                        Object::V0(o) => {
+                            for id in o.children {
+                                load_tree(nodes, missing, store, &id);
+                            }
+                        }
+                    }
+                }
+                Err(_) => missing.push(id.clone()),
+            }
+        }
+
+        let mut nodes: Vec<Object> = vec![];
+        let mut missing: Vec<ObjectId> = vec![];
+
+        load_tree(&mut nodes, &mut missing, store, &root_id);
+
+        if missing.is_empty() {
+            Ok(Tree {
+                root_id,
+                root_key,
+                nodes,
+            })
+        } else {
+            Err(missing)
+        }
+    }
+
+    /// Save objects of the tree in the store
+    pub fn save(&self, store: &Store) {
+        for node in &self.nodes {
+            store.put(node);
+        }
+    }
+
+    /// Get the ID of the root object
+    pub fn root_id(&self) -> ObjectId {
+        self.root_id
+    }
+
+    /// Get the key of the root object
+    pub fn root_key(&self) -> Option<SymKey> {
+        self.root_key
+    }
+
+    pub fn nodes(&self) -> Vec<Object> {
+        self.nodes
```

Comment:
does not compile:
cannot move out of `self.nodes` which is behind a shared reference
move occurs because `self.nodes` has type `Vec<types::Object>`, which does not implement the `Copy` trait

Reply:
I added a .clone() for now